### DISCUSSION TITLE
keep getting No response from time to time

### DISCUSSION THREAD

**Discussion by panosGkrigkas** [HomeSpan/HomeSpan#848] [Q&A]
I have uploaded a simple script to an esp32 which during testing it was working great. When I put the divice to it's final location I keep getting No response every 2-3 minutes for about 10 minutes and then the device is comming back online. I can ping the device and since a physical button is also attached, I can also control it from the button. The script is the following

relay.ino
```
#include "HomeSpan.h" 
#include "DEV_RELAY.h"

void setup() {
  Serial.begin(115200);

  homeSpan.enableWebLog(10,"pool.ntp.org","UTC","myLog");
  
  homeSpan.enableOTA();
  
  homeSpan.setPairingCode("xxxxxxxx");
  homeSpan.setQRID("xxx-xx-xxx");

  homeSpan.begin(Category::Switches, "Heater");

  // Accessory 2: Relay Switch
  new SpanAccessory();
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
    new DEV_RELAY(7, 2); // Initialize DEV_RELAY with relayPin 7 and buttonPin 2
}

void loop() {
  homeSpan.poll();
}
```
DEV_RELAY.h
```
struct DEV_RELAY : Service::Switch {
  int relayPin;
  int buttonPin;
  SpanCharacteristic *power;
  unsigned long startTime;
  unsigned long runTime = 3600000;

  DEV_RELAY(int relayPin, int buttonPin) : Service::Switch() {
    power = new Characteristic::On();
    new SpanButton(buttonPin); // Initialize SpanButton for the physical button

    this->relayPin = relayPin;
    this->buttonPin = buttonPin;
    pinMode(relayPin, OUTPUT); // Set relay pin as output
  }

  boolean update() {
    digitalWrite(relayPin, power->getNewVal()); // Update relay state based on power characteristic
    startTime = millis(); // Update start time
    return true;
  }

  void button(int pin, int pressType) override {
    if (pin == buttonPin) { // Check if the button pin is pressed
      if (pressType == SpanButton::LONG) { // Check for long press
        power->setVal(!power->getVal()); // Toggle power state on long press
        digitalWrite(relayPin, power->getNewVal()); // Update relay state based on power characteristic
        startTime = millis();
      }
    }
  }

  void loop() {
    if (power->getVal() && millis() - startTime >= runTime) { // Check if relay is ON and 10 seconds have elapsed
      power->setVal(false); // Turn off relay
      digitalWrite(relayPin, LOW);
      startTime = 0;
    }
  }
};
```

This is the log output
![image](https://github.com/HomeSpan/HomeSpan/assets/51501369/fdc9dae5-922d-4c32-a63f-f5ef7d879e8a)

**Comment by HomeSpan** [comment]
When the device is not responding to the Home App, does it still respond to weblog requests?

**Reply by panosGkrigkas** [reply]
Hi,

First of all thanks for the great work! Yes, the device does respond to weblog requests and shows no disconnection at all. The push button to locally control the device also runs normally. I think it has to do something with the way it interacts with my home pod mini serving as a bridge.

**Reply by HomeSpan** [reply]
This is likely due to the MDNS problem that seems to plague different configurations in different ways.  I assume you are using the Home Pod Mini as a HomeKit Hub? If so, when the device shows no response in the Home App, can you check to see if any sockets are connected to the device by typing 's' into the Serial Monitor.  My guess is that it will show that none of the sockets are connected.

If that's true, can you try the following test:  Add the following line to the `setup()` portion of your sketch:

```C++
new SpanUserCommand('D', " - disconnect WiFi", [](const char *buf){WiFi.disconnect();}); 
```

This allows you to force a WiFi disconnect by typing `@D` into the Serial Monitor.  When you start receiving a no-response error from the Home App, type `@D`.  HomeSpan will disconnect from WiFi, and then automatically re-connect after a second or two.  What we want to test is whether forcing a WiFi re-connect cures the MDNS problem.

**Reply by panosGkrigkas** [reply]
Thanks for the quick response and help, will try it out and revert one I complete this step

**Reply by phamvanan88** [reply]
First of all, thank you for the great library — it’s been working very reliably in my project.

I noticed that when typing s in the Serial Monitor, HomeSpan prints the list of active sockets / connections, which is very helpful for debugging cases where the Home app shows “No Response” even though Wi-Fi and local logic are still running.

I’d like to ask a clarification question:

Is the socket / connection count that is shown by the s command accessible programmatically from user code?

Or is it intentionally exposed only via the Serial Monitor for manual debugging?

My goal is to implement a very conservative, non-intrusive recovery mechanism (e.g. forcing a Wi-Fi reconnect only after many hours with zero HomeKit connections), but I want to do this only using supported and stable APIs.

If there is a recommended HomeSpan-native way to observe HomeKit connectivity status (even indirectly), I’d appreciate any guidance.

Thanks again for maintaining HomeSpan and for your time.

**Reply by HomeSpan** [reply]
There is currently no function call to return the number of active socket connections (i.e. the number in the 's' command), but I think this is a good idea.  Let me give some thought into the best way add an API for some form set of global "status" flags that could be monitored by the user directly in a sketch (maybe include whether WiFi is connected, Ethernet is connected, number of sockets, whether device is paired, etc.).  Could also add some logic to reboot directly if WiFi is on, pairing is on, and there has been no socket connectivity for a specified amount of time.

**Reply by phamvanan88** [reply]
Thanks for the explanation.
In my real usage, I often see “No Response” in the Home app while the device is still connected to Wi-Fi and running normally. This makes it hard to detect the problem from sketch code, and it seems likely related to mDNS or HomeKit session issues, not Wi-Fi itself.
If HomeSpan could expose some read-only status flags (Wi-Fi/paired state, socket activity, etc.), it would make it much easier to implement safe and conservative recovery logic (e.g. a Wi-Fi reconnect instead of a reboot).
Thanks again for considering this and for maintaining HomeSpan.

**Comment by Xorlent** [comment]
I exposed some resource info for you in this [pull request](https://github.com/HomeSpan/HomeSpan/pull/1147) if you want to see if it helps with your debugging.  You should be able to extend it to include other info as needed, or removing what you don't want (like PSRAM heap info).

**Comment by HomeSpan** [comment]
I've added a number of new methods and features to address the above issues:

1) The Web Log now includes a separate Client Connections table as an analog to the 's' CLI command allowing you to see if there are secure HomeKit connections (it also shows any ordinary connections, such as the one to the Web Log itself).
2) The state HS_PAIRED now reflects  the condition where HomeSpan is paired and connected to a network, but does NOT have any active secure connections to HomeKit.  The new state HS_CONNECTED has been added to indicate HomeSpan does have active connections to HomeKit.
3) New methods allow you to get the HomeSpan status and the duration of the status so you can programmatically determine if there are active HomeKit connections, or if for whatever reason they have all closed (or never opened) allowing you to take potential corrective action, such as reboot the device, disconnect/reconnect WiFi, etc., which I think should address the specific use-case you noted above.

All code is available in dev branch along with updated documentation on the new features and methods.  Please feel free to test and let me know how it goes. I'll be promoting to production version 2.8.1 in the near future if everything is working.

**Comment by HomeSpan** [comment]
This has now been released into production (HomeSpan 2.1.8).

**Comment by frankonski** [comment]
While I didn't notice the "No response" from Homekit, I have been experiencing constant connection issues when using Chrome to connect to a separate webserver (`httpd_handle_t`). Chrome kept erroring out with some kind of network / DNS issue. In my setup, I start this webserver instance when the `setConnectionCallback()` method is called (when HomeSpan connects to the WiFi).

One thing to note: My local domain name is `.local`: I run my own local DNS server serving `.local`. 

I noticed that when the board booted up, even after the WiFi was connected, no MDNS packets were ever sent, even though this is what's supposed to happen with ESPmDNS.

I was able to get this initial MDNS broadcast sent on power up by adding the following in the `setConnectionCallback()` method:
```
    esp_wifi_set_ps(WIFI_PS_NONE); 
    MDNS.addService("http", "tcp", "80"); // My webserver runs on port 80, while HAP is on 82.
  
    // Add a TXT record to force a broadcast of the IP address
    mdns_service_txt_item_set("_http", "_tcp", "version", "1.0");
```

Only with the above was I able to see the board broadcast its MDNS config on WiFi connect.
And that also resolved Chrome's connection issues to the web server on port 80.

I'm running HS 1.8.1 / IDF 5.5.4 / Arduino 3.3.9

**Reply by frankonski** [reply]
See https://github.com/espressif/esp-idf/issues/3121 regarding `esp_wifi_set_ps(WIFI_PS_NONE);`
